### PR TITLE
[export] Fix serialization for auto_functionalization

### PR DESCRIPTION
**Pull request by angelayi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118810

- Added support for serializig the auto_functionalization op, which
  required adding the functions `serialize_arbitrary_inputs` and
  `serialize_arbitrary_outputs` which will serialize the inputs/outputs
  without needing a schema, since HOOs do not have a schema.
- Added support for serializing user input mutations
- Added support for serializing operator inputs. They just get turned
  into strings.

Differential Revision: [D53331039](https://our.internmc.facebook.com/intern/diff/D53331039)

### CODE REVIEW

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D53331039).

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D53331039).

File: torch/_export/serde/serialize.py [hunk 1]
```diff
@@ -939,6 +948,57 @@ def output_node_at_index(node, index):
...
+            # e.g "-> Tensor"
+            return [Argument.create(as_tensor=self.serialize_tensor_output(node.name, meta_val))]
+        elif isinstance(meta_val, torch.SymInt):
+            # e.g "-> SymInt"
+            return [Argument.create(as_sym_int=self.serialize_sym_int_output(node.name, meta_val))]
+        elif isinstance(meta_val, torch.SymBool):
+            # e.g "-> SymBool"
+            return [Argument.create(as_sym_bool=self.serialize_sym_bool_output(node.name, meta_val))]
+
+        # There are a two possibilities at this point:
```
Comment (suo) [inline torch/_export/serde/serialize.py:968]: This implies that the only allowed outputs of a HOO are:
- Tensor
- Symint
- SymBool
- List[Tensor]
- Tuple[Tensor]

Is that right? Should we check and error on this explicitly? Right now I'm not sure what happens if we see a HOO that returns like `Tuple[Tensor, SymInt]`.

File: torch/_export/serde/serialize.py [hunk 2]
```diff
@@ -939,6 +948,57 @@ def output_node_at_index(node, index):
 
         return output_arguments
 
+    def serialize_arbitrary_outputs(self, node: torch.fx.Node) -> List[Argument]:
```
Comment (suo) [inline torch/_export/serde/serialize.py:951]: "arbitrary" doesn't seem right, given that we do not accept arbitrary outputs. Maybe just `serialize_hop_outputs` or whatever

**Review by suo — APPROVED** [review]

File: torch/_export/serde/serialize.py (hunk 1 above)
Comment (angelayi) [reply torch/_export/serde/serialize.py:968]: Added an assert, although I think I should fix it so that we can return Tuple[Tensor, SymInt]

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D53331039).

**Pull request closed without merging**